pkg/controller/githubsource/reconcile.go

```diff
@@ -0,0 +1,274 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package githubsource
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+
+	ghclient "github.com/google/go-github/github"
+	sourcesv1alpha1 "github.com/knative/eventing-sources/pkg/apis/sources/v1alpha1"
+	"github.com/knative/eventing-sources/pkg/controller/githubsource/resources"
+	"github.com/knative/eventing-sources/pkg/controller/sinks"
+	"github.com/knative/pkg/logging"
+	servingv1alpha1 "github.com/knative/serving/pkg/apis/serving/v1alpha1"
+	"go.uber.org/zap"
+	"golang.org/x/oauth2"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// reconciler reconciles a GitHubSource object
+type reconciler struct {
+	client              client.Client
+	scheme              *runtime.Scheme
+	dynamicClient       dynamic.Interface
+	recorder            record.EventRecorder
+	receiveAdapterImage string
+}
+
+// Reconcile reads that state of the cluster for a GitHubSource
+// object and makes changes based on the state read and what is in the
+// GitHubSource.Spec
+func (r *reconciler) Reconcile(ctx context.Context, object runtime.Object) (runtime.Object, error) {
+	logger := logging.FromContext(ctx)
+
+	source, ok := object.(*sourcesv1alpha1.GitHubSource)
+	if !ok {
+		logger.Errorf("could not find github source %v\n", object)
+		return object, nil
+	}
+
+	// See if the source has been deleted
+	accessor, err := meta.Accessor(source)
+	if err != nil {
+		logger.Warnf("Failed to get metadata accessor: %s", zap.Error(err))
+		return object, err
+	}
+	// No need to reconcile if the source has been marked for deletion.
+	deletionTimestamp := accessor.GetDeletionTimestamp()
+	if deletionTimestamp != nil {
+		return object, nil
+	}
+
+	reconcileErr := r.reconcile(ctx, source)
+	return source, reconcileErr
+}
+
+func (r *reconciler) reconcile(ctx context.Context, source *sourcesv1alpha1.GitHubSource) error {
+	source.Status.InitializeConditions()
+
+	if source.Spec.Repository == "" {
```

Comment:
I had the same thought. Do any of the other event sources do any kind of validation via a validating webhook today? I wasn't sure if we wanted to start a pattern of a new validating webhook for every source.

Reply:
I think they should. Ideally these would be broken into a single:
 CRD
 Controller
 Webhook

So that you can deploy a source as a single entity. Currently they all get baked into the same controller which is wonky, but works for now. Plans are to break them apart later.

Reply:
As much as possible the validation is happening using OpenAPI inside the CRD. There are a few examples of this in the container source.

Reply:
Because of the simple nature of this validation, I moved everything into kubebuilder annotations which feed into the OpenAPI schema.